### PR TITLE
aptos tutorial

### PR DESCRIPTION
**Pull request by mkurnikov**

### CODE REVIEW

File: tutorials/aptos-tutorial.md [hunk 1]
```diff
@@ -0,0 +1,271 @@
+# Aptos tutorial
+
+Aptos is a Layer 1 blockchain that allows developers to create their smart-contracts in Move language. 
+
+TODO: expand on Aptos
+
+Move is a unique smart-contract language, which was created with a heavy focus on security. Inspired by Rust, 
+it has a lot of restrictions which prevent developers from shooting themselves in a foot and miss
```
Comment (borispovod) [inline tutorials/aptos-tutorial.md:8]: let's replace restrictions with improvements?

File: tutorials/aptos-tutorial.md [hunk 2]
```diff
@@ -0,0 +1,271 @@
+# Aptos tutorial
+
+Aptos is a Layer 1 blockchain that allows developers to create their smart-contracts in Move language. 
+
+TODO: expand on Aptos
```
Comment (borispovod) [inline tutorials/aptos-tutorial.md:5]: Asked Alejo to helo here

File: tutorials/aptos-tutorial.md [hunk 3]
```diff
@@ -0,0 +1,271 @@
+# Aptos tutorial
+
+Aptos is a Layer 1 blockchain that allows developers to create their smart-contracts in Move language. 
+
+Aptos is focused on delivering the safest, fastest and most reliable production-ready Layer 1 blockchain to create creating universal and fair access to web3 for billions of people. Aptos recently announced a $200M Seed roundled by a16z crypto with participation from Multicoin Capital, Katie Haun, 3 Arrows Capital, ParaFi Capital and Coinbase Ventures, among others to scale the Aptos ecosystem. The team is comprised of the original creators, researchers, designers, and builders of Diem, the blockchain that was first built to serve this purpose. For more information on Aptos, please visit https://aptoslabs.com
```
Comment (borispovod) [inline tutorials/aptos-tutorial.md:5]: Let's make it's some kind of quote

File: tutorials/aptos-tutorial.md [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+
+Move projects are called Packages, and contain multiple Modules. 
+
+Module is a smart-contract that combine types and functions, and provides a very unique set of rules and restrictions 
+to manage relationships between them.  
+
+Package is a set of related Move modules which share common dependencies and often published
+to the same address in the Aptos blockchain.
+
+First, create a directory and initialize it with a Package using aptos-cli:
```
Comment (borispovod) [inline tutorials/aptos-tutorial.md:50]: using aptos binary?

File: tutorials/aptos-tutorial.md [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+`AptosFramework` - a set of modules specific to Aptos blockchain, like `Coin` module for ERC20-like, and `Account` 
+for the account metadata. 
+
+### Resources and storage
+
+TODO: talk about &signer and resources
+
+Every user of the Aptos blockchain has it's own object storage, located at the user address. There are builtin methods, 
+which allow to communicate with said storage from the Move code.
```
Comment (borispovod) [inline tutorials/aptos-tutorial.md:125]: Maybe let's mention that arguments for that functions usually addresses, but to store something dev needs signer?

File: tutorials/aptos-tutorial.md [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+We also add `acquires InfoStore` explained before. 
+
+```move
+module Sender::UserInfo {
+    use Std::ASCII::String;
+    use Std::Signer;
+    
+    struct InfoStore has key { username: String }
+    
+    public fun get_username(account: &signer): String acquires InfoStore {
```
Comment (borispovod) [inline tutorials/aptos-tutorial.md:190]: why we need signer here? if i want to look at someone else username?

File: tutorials/aptos-tutorial.md [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+        let account_addr = Signer::address_of(account);
+        borrow_global<InfoStore>(account_addr);
+    }
+  
+    public fun set_username(account: &signer, username: String) acquires InfoStore {
```
Comment (borispovod) [inline tutorials/aptos-tutorial.md:213]: in the next part we should make it script

File: tutorials/aptos-tutorial.md [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+### Tests
+
+Now, let's write a test for our module to make sure everything works correctly. Test functions could be added anywhere, 
+but it's a nice convention to store those in the `tests/{$MODULE_NAME}Tests.move` module. 
+
+Create a `tests/` directory to the root of the package, and add `UserInfoTests.move` there. Test functions are marked 
+with `#[test]` attribute, this way Move knows it's a test and can apply special treatment:
+
+```move
+module Sender::UserInfoTests {
```
Comment (borispovod) [inline tutorials/aptos-tutorial.md:237]: You forgot #[test_only] before module :)

File: tutorials/aptos-tutorial.md [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+        let username = ASCII::string(b"MyUser");
+        UserInfo::set_username(&account, username);
+      
+        // assert! macro for asserts, needs an expression and a failure error code 
+        assert!(UserInfo::get_username(&account) == username, 1);
+    }          
+}
+```
+
+Now, to run this test:
```
Comment (borispovod) [inline tutorials/aptos-tutorial.md:256]: Would be good to continue with:

* Convert public function to script
* Publish modules
* Run scripts

I can help with it if needed

**Review by borispovod — CHANGES_REQUESTED** [review]

File: tutorials/aptos-tutorial.md [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+it has a lot of restrictions which prevent developers from shooting themselves in a foot and miss 
+the critical vulnerability.
+
+In this tutorial, we're going to create a simple module that allows users to store their username in the blockchain. 
+This will allows us to explore unique features of Move and prepare for the more serious projects.
+
+### Aptos CLI
+
+First we need to install Aptos CLI. Go to the Aptos CLI releases page at 
+[https://github.com/aptos-labs/aptos-core/releases], and download zip archive for your OS.
```
Comment (borispovod) [inline tutorials/aptos-tutorial.md:15]: broken link

File: tutorials/aptos-tutorial.md [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+```
+userinfo/
+├── sources/
+└── Move.toml
+```
+
+`sources/` - directory is where you put your modules.  
+`Move.toml` - manifest file for the package. Here, you define package metadata, dependencies and addresses used in the code.
+
+Let's add an address with name `Sender` and value `0x42` under the `[addresses]`. We're going to store all our modules at that
```
[the remark this answers is not in the document]
Comment (mkurnikov) [reply tutorials/aptos-tutorial.md:67]: That's just an arbitrary value

File: tutorials/aptos-tutorial.md [hunk 12]
```diff
@@ -278,3 +286,191 @@ Test result: OK. Total tests: 1; passed: 1; failed: 0
...
+No key given, generating key...
+
+Account 7EBD304E2E7E7C2147D14A605072480473256B75D5CC7A9A794C0C5BC2F1E17B doesn't exist, creating it and funding it with 10000 coins
+Aptos is now set up for account 7EBD304E2E7E7C2147D14A605072480473256B75D5CC7A9A794C0C5BC2F1E17B!  Run `aptos help` for more information about commands
+{
+  "Result": "Success"
+}
+```
+
+The deployment config which contains the private key of the new account would be created in `.aptos/config.yaml`. Don't share your private key with anyone! If you want to change configuration or account, just run `init` again, the configuration will be overwritten.
```
Comment (mkurnikov) [inline tutorials/aptos-tutorial.md:320]: `.aptos/config.yaml` in the root of the project?

**Pull request merged**